generate-offline-map-using-workmanager/src/main/java/com/esri/arcgismaps/sample/generateofflinemapusingworkmanager/OfflineJobWorker.kt

```diff
@@ -0,0 +1,153 @@
+        // load the json file
+        val offlineJobJsonFile = File(offlineJobJsonPath)
+        // if the file doesn't exist return failure
+        if (!offlineJobJsonFile.exists()) {
+            return Result.failure()
+        }
+        // create the GenerateOfflineMapJob from the json file
+        val generateOfflineMapJob =
+            GenerateOfflineMapJob.fromJson(offlineJobJsonFile.readText())
+            // return failure if the created job is null
+                ?: return Result.failure()
+
+        return try {
+            // set this worker to run as a long-running foreground service
+            // this will throw an exception, if the worker is launched when the app
+            // is not in foreground
+            setForeground(createForegroundInfo(0))
+            // A deferred to wait for the completion of the generateOfflineMapJob
+            val deferred = CompletableDeferred<Result>()
+            // check and delete if the offline map package file already exists
+            // this check is needed, if the download has failed midway and is restarted later
+            // by WorkManager
+            File(generateOfflineMapJob.downloadDirectoryPath).deleteRecursively()
+
+            // start the generateOfflineMapJob
+            // this job internally runs on a Dispatchers.IO context, hence this CoroutineWorker
+            // can be run on the default Dispatchers.Default context
+            generateOfflineMapJob.start()
+
+            // launch a job progress collector
+            coroutineScope.launch {
+                generateOfflineMapJob.progress.collect { progress ->
+                    // update the worker progress
+                    setProgress(workDataOf("Progress" to progress))
+                    // update the ongoing progress notification
+                    setForeground(createForegroundInfo(progress))
+                }
+            }
+
+            // launch a job status collector
+            coroutineScope.launch {
+                generateOfflineMapJob.status.collect { jobStatus ->
+                    if (jobStatus == JobStatus.Succeeded) {
+                        // if the job is successful show a final status notification
+                        workerNotification.showStatusNotification("Completed")
+                        // complete the deferred with a success result
+                        deferred.complete(Result.success())
+                    } else if (jobStatus == JobStatus.Failed) {
+                        // if the job has failed show a final status notification
+                        workerNotification.showStatusNotification("Failed")
+                        // complete the deferred with a failure result
+                        deferred.complete(Result.failure())
+                    }
+                }
+            }
+
+            // wait for the completion of the deferred value and return its result
+            deferred.await()
```

Comment:
No need to use a CompletableDeferred here. You have two options:
1. Use `Job.result()` here.
2. Rely on the use of the `coroutineScope` function as suggested above. `coroutineScope` will not resume until all of its child coroutines have completed. So within the `coroutineScope` block you could simply collect progress and status event and as soon as you get a job status of `Succeeded` or `Failed`, you terminate the collection of status and progress flows. You can do this by using predicates while collecting, for example [takeWhile](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/take-while.html).

Reply:
Thanks for the feedback! I've refactored it to use `Job.result()` and `takeWhile` and removed the use of a `CompletableDeferred`